### PR TITLE
Use .split('+') to separate plain version from git hash

### LINKED ISSUE #37: hideSha Option does not work for more complex Version numbers
For example the following are all valid [SemVer](http://semver.org/) Versions and they are not accepted by the helper when the hideSha Option is used. What is worse it breaks the whole app.

* 1.10.0 (Version number with more than one digit)
* 1.0.0-alpha.1 (Version number with a pre-release)

The second example doesn't actually raise an error but it is just stripped from the version (would return 1.0.0).

The reason being that the versionRegExp used doesn't match those cases: ```/\d[.]\d[.]\d/```

Option hideSha was introduced in #36 

I have several suggestions:

* Make the matching more robust. I don't even see a reason to use a regex. Splitting on + (the sign used in SemVer to separate build metadata from the rest of the version) should be enough.
* If for some reason something doesn't match return an empty string but don't break the whole app

As long as the hideSha Option is not used everything should continue to work as before so this is not critical. But especially the first issue (not accepting version 1.10.0) will probably lead to some head-scratching when using the Option.

### PR DESCRIPTION
**Pull request by omarkohl**
This is more robust and works correctly for versions such as 1.10.0 or
1.0.0-alpha.1 (which was not the case before).

Remove version regexes since they are no longer required.

closes #37 

@taras I was unable to figure out how to write tests with a predefined version number. Therefore right now tests/integration/helpers/app-version-test.js will fail. Any tips?

### CODE REVIEW

**Comment by taras** [comment]
Thank you for this! Can you create a unit test that imports the function from the helper and asserts on return value from the function?

**Comment by omarkohl** [comment]
There are already unit tests in tests/unit/helpers/app-version-test.js that still work.

But I have the same problem there.

I would rather test against predefined values to verify they are returned as expected instead of using the version number and git hash of ember-cli-app-version itself.

Since `version` is more or less a global variable (taken out of the config) I don't know how to override it.

The failing test can of course be modified to check something like "assert not empty" or "assert a certain length" or "assert certain character do or don't appear".

What do you think?

**Comment by taras** [comment]
Right, you could do something like this:

```js
test('it returns app version', function(assert) {
  assert.expect(3);

  assert.equal(appVersion('1.0.0'), '1.0.0');
  assert.equal(appVersion('1.0.0+abc', { hideSHA: true }), '1.0.0'); 
  assert.equal(appVersion('0.0.0+abc', { hideVersion: true }), 'abc');  
});
```

What do you think?

**Comment by omarkohl** [comment]
That would be nice but it doesn't work :-)

appVersion ignores the first parameter and uses the global version of the application taken out of the config. In case of the tests it's the `dummy` application (which seems to use the same version and git hash as the real ember-cli-app-version application).

Just one of the errors:

```
not ok 15 PhantomJS 2.1 - Unit | Helper | app version: it returns app version
    ---
        actual: >
            2.0.2+9ad6fe14
        expected: >
            1.0.0
        stack: >
            http://localhost:7357/assets/tests.js:213:17
            runTest@http://localhost:7357/assets/test-support.js:3257:32
            run@http://localhost:7357/assets/test-support.js:3242:11
            http://localhost:7357/assets/test-support.js:3384:14
            process@http://localhost:7357/assets/test-support.js:3043:24
            begin@http://localhost:7357/assets/test-support.js:3025:9
            http://localhost:7357/assets/test-support.js:3085:9
        Log: |
    ...
```

**Comment by taras** [comment]
My bad, you have to invoke the helper with the same arguments signature as a helper.

`appVersion(<array>params, <object>hash)`

`appVersion(['1.0.0'], { hideSHA: true })`

**Comment by omarkohl** [comment]
nope same error

File: tests/integration/helpers/app-version-test.js [hunk 1]
```diff
@@ -11,15 +10,13 @@ test('it displays only app version', function(assert) {
...
 });
 
 test('it displays only git sha', function(assert) {
   assert.expect(2);
 
   this.render(hbs`{{app-version hideVersion=true}}`);
 
-  assert.ok(this.$().text().match(shaRegExp), 'Displays git sha.');
-  assert.ok(!this.$().text().match(versionRegExp), 'Does not display version.');
+  assert.equal(this.$().text(), 'deadbeef', 'Displays git sha.');
```
Comment (alexlafroscia) [inline tests/integration/helpers/app-version-test.js]: Maybe I'm missing something, but why is this asserting that the text matches `deadbeef`?

**Comment by alexlafroscia** [comment]
The problem is that the helper should be stubbing the value of the configuration object, because as @omarkohl pointed out, it's using the version from the `config/environment` file.

To solve this, I would instead import the configuration by resolving `config:environment` off of the owner API, and then stubbing out the environment config in the resolver in each test.

The helper would end up looking something more like this (**warning:** this exact code probably won't work because I'm just writing off the top of my head)

```javascript
import Ember from 'ember';

export default Ember.Helper.extend({
  compute(_, hash = {}) {
    const version = Ember.getOwner(this).lookup('config:environment').APP.version;
    // ... Do the rest of the existing logic
  }
});
```

then your test can stub out `config:environment` the way a test would normal substitute something on the resolver.  An example of a test doing that can be found [here](https://gist.github.com/alisdair/00756de8c248bb5ced47cd2ea9e23b5a#file-tests-integration-components-spotify-song-test-js-L21).

**Comment by alexlafroscia** [comment]
The syntax that @taras suggested above assumes that you pass the app version into the helper, but that's not the case; the helper ignores any positional parameters, looks up the config directly by importing the environment file, and then only uses the named parameters (in the hash) to configure the output value.

**Comment by omarkohl** [comment]
@alexlafroscia Thanks for the pointers! I'll have a look.

File: tests/integration/helpers/app-version-test.js (hunk 1 above)
Comment (omarkohl) [inline tests/integration/helpers/app-version-test.js]: To make obvious what is missing. Just like in line 13 where I check for version 1.2.3
The values returned by the app currently are those of ember-cli-app-version which of course mean it is no possible to test for exact values.
But I'll have a look at your suggestions in the pull request on how to stub these values in the config and then test here for the stubbed value.

**Comment by taras** [comment]
@alexlafroscia you're right. Sorry about that. I should have looked closer but I had very busy few months - clearing up now.

@omarkohl if you refactor that function to accept a version and then use that function then it'll be possible to test it.

```js
import Ember from 'ember';
import config from '../config/environment';
import { shaRegExp, versionRegExp } from 'ember-cli-app-version/utils/regexp';

const {
  APP: {
    version
  }
} = config;

export function makeHelper(version) {
  return function appVersion(_, hash = {}) {
    if (hash.hideSha) {
      return version.match(versionRegExp)[0];
    }

    if (hash.hideVersion) {
      return version.match(shaRegExp)[0];
    }

    return version;
  }
}

export default Ember.Helper.helper(makeHelper(version));
```

In tests, you'll be able to do this,

```js
makeHelper('1.0.0')([], { hideSHA: true })
makeHelper('1.0.0+abc')([], { hideSHA: true })
```

@alexlafroscia what do you think?

**Comment by alexlafroscia** [comment]
@taras yeah, that looks like a good solution too! 👍 for a higher order function.

**Comment by taras** [comment]
@omarkohl will you be able to finish this or would you like me to pick it up?

**Comment by GottfridL** [comment]
Looking forward to get this fixed soon. I have a version number for example 0.1.137 that currently is cut to 0.1.1.

**Comment by omarkohl** [comment]
@alexlafroscia I would have preferred to implement something like what you suggested with:

```javascript
Ember.getOwner(this).lookup('config:environment').APP;
```

but I didn't manage to make it work.

Your code does not work out of the box and I tried several alternative such as these: https://discuss.emberjs.com/t/best-practices-accessing-app-config-from-addon-code/7006/20 but I was unable to figure it out.

I think it would be a better solution because it allows for 'end-to-end' testing ensuring that the version from the environment is actually used.

I implemented the solution @taras suggested with a closure.

**Comment by omarkohl** [comment]
@taras Please merge :-)

**Comment by elbeezy** [comment]
ping. I'm currently using @omarkohl's fork to use this feature, would like to see it merged if possible!

**Comment by Techn1x** [comment]
@taras let's get this merged :)

**Comment by alechirsch** [comment]
Any chance of getting this merged?

**Comment by Techn1x** [comment]
@rwjblue seems this PR was missed for the last release ?

**Comment by Techn1x** [comment]
FWIW I made a similar PR (https://github.com/ember-cli/ember-cli-app-version/pull/60), that essentially does the same but keeps the regex method and improves on it to allow for the cases outlined above.

It is also backwards compatible with existing flags, but adds new ones, so you can still just show the version number without (or with) the extended tag.

I would be happy if either of these PR's are merged in...................

**Pull request closed without merging**